### ISSUE TITLE
bug(rust/windows): TunnelDevice("Failed to create adapter")

### ISSUE THREAD

**Issue by ReactorScram** [CloudFire-LLC/cloudfire#6547]
# Describe the bug

Customer can't log in to Firezone because ~~connlib~~ Client is failing to create the adapter.

# Next steps

- [x] Get IPC service logs from customer
- [x] Gracefully sign out if we get a recoverable error while connecting, so that we can still export logs (#6548)
- [x] Send customer a debug MSI
- [ ] Debug IPC service

# Hypotheses

**Open**

- Maybe something is wrong with the permissions in the IPC service due to a group policy or MDM or something?
- Maybe an adapter leaked and so we failed to create a 2nd adapter with the same params?

**Proven**

- Logs could not be exported - Yes, #6548 will patch it

**Disproven**

- Maybe the IPC service is not running? - No, we got logs from it.
- Maybe Tailscale or Wireguard apps are interfering? - No, I signed in to those and Firezone could still open and connect just fine.

# To Reproduce

1. Sign in to Firezone
2. ???
3. Unable to raise tunnel
4. GUI shows error `Failed to connect to Firezone for non-Portal-related reason`
5. When exporting logs, GUI panics with `SendError` at `ctlr_tx.blocking_send(ControllerRequest::ExportLogs { path, stem }).unwrap()` because `run_controller` has already bailed out

# Expected behavior

Expected to be able to raise the tunnel

# Screenshots / Logs

(May post heavily redacted customer logs eventually)

# Platform (please complete the following information)

- _Component (i.e. macOS client / Linux client / Gateway / Admin portal):_ Windows Client
- _Firezone Version (e.g. 1.0.0 or N/A):_ 1.3.0
- _OS and version: (e.g. Ubuntu 22.04 or N/A):_ Windows unknown version
- _Deployment method: (e.g. Docker / Systemd / App Store or N/A):_ Official release MSI

# Additional context

Customer encountered this.

Bug is _not_ fixed in main, I've never seen this bug before.

**Comment by jamilbk** [comment]
Only two things come to mind as to what it could be:

- Some policy is preventing the creation of the adapter (is it possible the IPC service isn't running as admin?)
- There is already a VPN active, or some other conflicting software

**Comment by ReactorScram** [comment]
Yeah, something like that. The only place that error can be generated is this call to  `make_tun`:

https://github.com/firezone/firezone/blob/c66f0c15c0866b099ccb5e411d1b3228b06b6d97/rust/headless-client/src/ipc_service.rs#L517-L520

Which in turn is likely failing here:

https://github.com/firezone/firezone/blob/c66f0c15c0866b099ccb5e411d1b3228b06b6d97/rust/bin-shared/src/tun_device_manager/windows.rs#L229

Hopefully we'll see one of those error variants when we get the IPC service logs

**Comment by ReactorScram** [comment]
Customer reports the issue fixed in df3067a8f8, but we didn't merge anything near `Adapter::create` that should fix it. I'll keep this open a little longer.

**Comment by jamilbk** [comment]
Possibly related support request via HubSpot (still awaiting logs)

```
I've got this error : Failed to connect to Firezone error=TunnelDevice("Élément introuvable. (0x80070490)")
```

**Comment by ReactorScram** [comment]
Looks like it says "Element could not be found"
The error code may not be present for the customer we originally opened this ticket for, since we weren't printing detailed error codes at that time.

**Comment by Intuinewin** [comment]
About https://github.com/firezone/firezone/issues/6547#issuecomment-2346862245

I spent too much time trying to debug this, it seems we have this issue because IPv6 is disabled on our computers

Line that produces this error:
https://github.com/firezone/firezone/blob/4e821a3c30514c59289b8acc6e22a36b721a9b6c/rust/bin-shared/src/tun_device_manager/windows.rs#L400

To reproduce:
```
reg add "HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Tcpip6\Parameters" /v DisabledComponents /t REG_DWORD /d 255 /f
```

Sidenote: on 1.3.2, we don't have the GUI error, the IPC service fails silently

**Comment by thomaseizinger** [comment]
> IPv6 is disabled on our computers

This makes sense! I think instead of `?`, in both cases (IPv4 and IPv6) we should use an `if let Some()` and only update the interface entry if it actually exists.

**Comment by thomaseizinger** [comment]
Thank you for debugging this @Intuinewin ! :)

**Comment by ReactorScram** [comment]
Hm, I couldn't replicate it with 1.3.2 and that registry command. Neither on bare metal x86_64 nor aarch64 VM guest.

**Comment by Intuinewin** [comment]
Have you restarted your device after the registry update? It's weird I can replicate this on my laptop at home

**Comment by ReactorScram** [comment]
Okay yeah, if I restart after adding the registry value, it replicates for me. Both on 1.3.2 and 35ffc99b11 (tip of main)

**Comment by ReactorScram** [comment]
@Intuinewin Thanks for the report, I tested #6681 locally and that should solve it. I'm keeping this issue open because I'm pretty sure the other customer hit a different error

**Comment by Intuinewin** [comment]
Thanks for the fix, I've tested #6681 and it works on my machine :tm:

**Comment by thomaseizinger** [comment]
Closing this because it is not actionable. There can be many reasons why we can't create the adapter and we now have better logging and reporting in place that tells us why.
